### PR TITLE
Improve Heap extraction performance

### PR DESCRIPTION
**Pull request by RushiT0122**
# Description of change
- Improve the extraction performance by adding multi-processing.
- Improve logging

# Manual QA steps
 - Tested changes on the cloned connection.
 - Conducted performance testing to determine the optimal number of threads.
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_heap/sync.py [hunk 1]
```diff
@@ -68,13 +86,27 @@ def get_files_to_sync(table_manifests, table_name, state, bucket):
                     for file_name in manifest['files']], key=key_fn)
 
     if bookmark and bookmarked_version and bookmark in files:
-        #NB> The bookmark is a fully synced file, so start immediately
-        #after the bookmark
+        # NB> The bookmark is a fully synced file, so start immediately
+        # after the bookmark
         files = files[files.index(bookmark)+1:]
 
+    LOGGER.info("Found %d menifest files.", len(files))
```
Comment (longestpath) [inline tap_heap/sync.py:93]:
```suggestion
    LOGGER.info("Found %d manifest files.", len(files))
```

File: tap_heap/sync.py [hunk 2]
```diff
@@ -1,15 +1,33 @@
...
 
 LOGGER = singer.get_logger()
 
+# Define global queue
+QUEUE_TIMEOUT = 120
+QUEUE_MAX_LIMIT = 20000
+record_queue = multiprocessing.Queue(maxsize=QUEUE_MAX_LIMIT)
+
+
+# This event will signal all producer and consumer threads to stop thier execution
```
Comment (longestpath) [inline tap_heap/sync.py:26]:
```suggestion
# This event will signal all producer and consumer threads to stop their execution
```

File: tap_heap/sync.py [hunk 3]
```diff
@@ -68,13 +86,27 @@ def get_files_to_sync(table_manifests, table_name, state, bucket):
...
+    # wait for queue to get empty after terminate event is set.
+    while not terminate_event.is_set() or record_queue.qsize():
+        try:
+            singer.write_message(record_queue.get(timeout=5))
+        except queue.Empty:
+            continue
+        except Exception as ex:    # pylint: disable=broad-exception-caught
+            terminate_event.set()
+            raise ProcessError("Consumer thread stopped abruptly!") from ex
+    LOGGER.info("Existing from the consumer thread!")
```
Comment (longestpath) [inline tap_heap/sync.py:107]:
```suggestion
    LOGGER.info("Exiting from the consumer thread!")
```

File: tap_heap/sync.py [hunk 4]
```diff
@@ -98,42 +130,96 @@ def sync_stream(bucket, state, stream, manifests):
...
+                    records_streamed += future.result()
+                except Exception as ex:     # pylint: disable=broad-exception-caught
+                    stored_exception = ex
+                    terminate_event.set()
+                    break
+
+            if terminate_event.is_set():
+                raise Exception(f"Error reading file {file_path}") from stored_exception     # pylint: disable=broad-exception-raised
+
+            LOGGER.info("Extracted %d/%d menifest files.", i + batch_size, len(files))
```
Comment (longestpath) [inline tap_heap/sync.py:157]:
```suggestion
            LOGGER.info("Extracted %d/%d manifest files.", i + batch_size, len(files))
```

File: tap_heap/sync.py [hunk 5]
```diff
@@ -98,42 +130,96 @@ def sync_stream(bucket, state, stream, manifests):
...
+                                        key_properties=key_properties),
+                timeout=QUEUE_TIMEOUT)
+
+        records_synced = 0
+        with Transformer() as transformer:
+            for row in iterator:
+                # Terminate the thread execution
+                # if any of produceror consumer threads exits abruptly
+                if terminate_event.is_set():
+                    raise ProcessError("Recieved event to terminate the thread abruptly!")
```
Comment (longestpath) [inline tap_heap/sync.py:210]:
```suggestion
                    raise ProcessError("Received event to terminate the thread abruptly!")
```

**Review by longestpath — CHANGES_REQUESTED** [review]
Just some typos, seems solid as long as it was proven in practice

File: tap_heap/sync.py (hunk 1 above)
Comment (RushiT0122) [reply tap_heap/sync.py]: Fixed

File: tap_heap/sync.py (hunk 2 above)
Comment (RushiT0122) [reply tap_heap/sync.py]: Fixed

File: tap_heap/sync.py (hunk 3 above)
Comment (RushiT0122) [reply tap_heap/sync.py]: Fixed

File: tap_heap/sync.py (hunk 4 above)
Comment (RushiT0122) [reply tap_heap/sync.py]: Fixed

File: tap_heap/sync.py (hunk 5 above)
Comment (RushiT0122) [reply tap_heap/sync.py]: Fixed

**Review by dmosorast — CHANGES_REQUESTED** [review]
Would it be possible to get the tests passing before merging? It's been awhile since a change was made to this repo, so it'd be good to make sure the unit tests still pass.

**Review by dmosorast — APPROVED** [review]

**Pull request merged**